### PR TITLE
[core] Add cgroups support to meshroom

### PR DESCRIPTION
**Pull request by servantftechnicolor**
CGroups on linux may limit the memory and the number of cpus

Without getting the specific information about these limits, we may overshoot our memory usage and core usage, creating crashes.

This PR will read the values as specified by the specifications.

Rely on https://github.com/alicevision/AliceVision/pull/1304.

### CODE REVIEW

**Comment by fabiencastan** [comment]
We need to find a way for nodes to declare if they supports these arguments or rely on environment variables.
We have many nodes launching external commands (like blender).
We could for instance create an intermediate class for AVCommandLineNode.

**Comment by mh0g** [comment]
we could use a  chunk.node.nodeDesc.commandLine.startswith("aliceVision_") to do this only on the proper nodes

**Review by mh0g — APPROVED** [review]

**Comment by mh0g** [comment]
a warning in the log when the feature is used may be usefull

File: meshroom/core/cgroup.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        return -1
+
+    size = -1
+    filename = f"/sys/fs/cgroup/memory/{cgroup}/memory.limit_in_bytes"
+    try:
+        with open(filename, "r") as f :
+            value = f.read().rstrip("\r\n")
+            if value.isnumeric():
+                size = int(value)
+    except IOError:
```
Comment (fabiencastan) [inline meshroom/core/cgroup.py:41]: OSError

File: meshroom/core/desc.py [hunk 2]
```diff
@@ -589,6 +593,23 @@ def processChunk(self, chunk):
...
             chunk.subprocess = None
 
+#specific command line node for alicevision apps
+class AVCommandLineNode(CommandLineNode):
+    def buildCommandLine(self, chunk):
+
+        str = super(AVCommandLineNode, self).buildCommandLine(chunk)
+
+        cmdMem = ''
+        memSize = cgroup.getCgroupMemorySize()
```
Comment (fabiencastan) [inline meshroom/core/desc.py:603]: This is a static information that should be estimated once for the lifetime of the session and not each time we evaluate the command line per node.

**Review by fabiencastan — CHANGES_REQUESTED** [review]

File: meshroom/core/cgroup.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        with open(filename, "r") as f :
+            value = f.read().rstrip("\r\n")
+            if value.isnumeric():
+                size = int(value)
+    except OSError:
+        pass
+
+    return size
+
+def parseNumericList(str):
```
Comment (fabiencastan) [inline meshroom/core/cgroup.py:46]: str is a python keyword, it would be good to rename this variable.

File: meshroom/core/desc.py [hunk 4]
```diff
@@ -589,6 +593,34 @@ def processChunk(self, chunk):
...
+            AVCommandLineNode.cmdCore = ''
+            coresCount = cgroup.getCgroupCpuCount()
+            if coresCount > 0:
+                AVCommandLineNode.cmdCore = ' --maxCores={coresCount}'.format(coresCount=coresCount)
+
+            AVCommandLineNode.cgroupParsed = True
+
+    def buildCommandLine(self, chunk):
+
+        str = super(AVCommandLineNode, self).buildCommandLine(chunk)
```
Comment (fabiencastan) [inline meshroom/core/desc.py:621]: same

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**